### ISSUE TITLE
Binary operator with return type matching upper bound of given TypeVar causes type checking failure

### ISSUE THREAD

**Issue by jfmcoronel** [microsoft/pyright#2472]
**Describe the bug**
Pyright 1.1.180 fails to do correct type checking of values resulting from the use of a built-in binary operator (e.g., `+`, `**`) with one of its operands annotated with a type variable having `int` as its upper bound. mypy 0.910 flags this case with an error. The same issue is exhibited for `str` and `+`.

This seems to be due to the fixed type of the binary operator's return value (e.g., `int` for `x + 1`) being seen as compatible with the upper bound of the type variable. This does not seem to be correct as the actual type of the type variable can be any subtype of the upper bound (not just the upper bound itself).

**To reproduce**
Run the code below and verify that it erroneously passes Pyright 1.1.180 checking.

**Expected behavior**
Pyright should flag `return x` with an error message similar to `Incompatible return value type (got "int", expected "T")` of mypy 0.910.

**Screenshots or code**

```python
# pyright: strict

from typing import TypeVar

T = TypeVar('T', bound=int)

def f(n: T) -> T:
    x = n + 1       # Reproducible with +, -, *, /, **, %, |, ^, and &
    reveal_type(x)  # Pyright: Type of "x" is "int*"; mypy: Revealed type is "builtins.int"

    return x        # mypy: Incompatible return value type (got "int", expected "T")
                    # Pyright: Passes type checking
f(1)
```

Using strings instead of integers and `+` as the binary operator in the above snippet also seems to trigger this bug:

```python
# pyright: strict

from typing import TypeVar

T = TypeVar('T', bound=str)

def f(n: T) -> T:
    x = n + 'a'
    reveal_type(x)  # Pyright: Type of "x" is "str*"; mypy: Revealed type is "builtins.str"

    return x        # mypy: Incompatible return value type (got "str", expected "T")
                    # Pyright: Passes type checking
f('a')
```

**VS Code extension or command-line**
Checked via Pyright command-line tool, coc.nvim of Vim, and Pylance of VS Code.

**Additional context**

The bug seems to be caused only by binary operators; using the unary operator `~` (e.g., `x = ~n`) results in Pyright correctly flagging the error.

Annotating `x` above incorrectly with `T` (e.g., `x: T = n + 1`) also seems to be overlooked by Pyright; mypy flags this as well.
